### PR TITLE
RATIS-940. Add sonar check for ratis

### PR DESCRIPTION
**Pull request by dineshchitlangia**
Ratis instance has been created in ASF's Sonar Cloud. This can now be used to analyze and improve the code quality. To ensure the analysis is automatically executed whenever there is a commit in master branch, we need to add a github action for Sonar check.
This PR aims to do that.

https://issues.apache.org/jira/browse/RATIS-940

For testing, I was able to add a token in my fork on incubator-ratis and confirm the script triggers analysis. For the incubator-ratis repo, ASF INFRA team helped to add the sonar cloud token in the repo settings via https://issues.apache.org/jira/browse/INFRA-20295
Once this PR is merged, I can confirm that the sonar check got enabled correctly.

### CODE REVIEW

**Comment by dineshchitlangia** [comment]
@elek Could I request your quick review as you had done the similar changes in Ozone?
Unlike your changes in Ozone, I am performing the sonar check first and then the unit tests as Ratis has some broken UTs. Those will ensure the unit check part exits out once it fails the unit tests and may not trigger the sonar check.

**Comment by dineshchitlangia** [comment]
Checkstyle failure is unrelated to the PR.

ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/LogSegment.java
 47: Unused import - org.apache.ratis.server.raftlog.RaftLog.INVALID_LOG_INDEX.

File: dev-support/checks/sonar.sh [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd)"
+cd "$DIR/../.." || exit 1
+
+if [ ! "$SONAR_TOKEN" ]; then
+  echo "SONAR_TOKEN environment variable should be set"
+  exit 1
+fi
+mvn -B verify -DskipShade -DskipTests -Dskip.yarn org.sonarsource.scanner.maven:sonar-maven-plugin:3.6.0.1398:sonar -Dsonar.host.url=https://sonarcloud.io -Dsonar.organization=apache -Dsonar.projectKey=apache_incubator-ratis
```
Comment (mukul1987) [inline dev-support/checks/sonar.sh:23]: skip.yarn is not needed as this is not hadoop code

Comment (dineshchitlangia) [reply dev-support/checks/sonar.sh:23]: @mukul1987 I think you meant skipShade is not needed. I will update the PR.

**Comment by dineshchitlangia** [comment]
@elek Thank you for the review. You are right the Github token is not needed when merging/pushing locally. As we are slowly moving towards merging directly via GitHub PRs this will be needed, at least as per Sonar's documentation :)

Let me know if you can commit this or you want me to drop the github token.

**Comment by elek** [comment]
> this will be needed, at least as per Sonar's documentation :)

Ok, thanks. Good to know. In that case, let me merge it.

**Comment by dineshchitlangia** [comment]
Thanks @elek  for review/commit.
Thanks @mukul1987  for reviews.

I confirmed we are able to see the Sonar analyses : https://sonarcloud.io/dashboard?id=apache_incubator-ratis

I will create an action plan / umbrella Jira to fix all of these and we should be in a clean state soon.

**Comment by elek** [comment]
> I confirmed we are able to see the Sonar analyses : https://sonarcloud.io/dashboard?id=apache_incubator-ratis

Wow. Great work.

**Pull request merged**